fraim/outputs/sarif.py

```diff
@@ -169,15 +207,97 @@ class SarifReport(BaseSchema):
+                        execution_successful=True,
+                        toolExecutionNotifications=[
+                            Notification(
+                                level="error",
+                                descriptor=ReportingDescriptorReference(id="PARSING_ERROR"),
+                                message=Message(
+                                    text=f"Code chunk could not be analyzed due to a parsing error: {failure.reason}"
+                                ),
+                                locations=failure.chunk.locations.to_sarif(),
+                            )
+                            for failure in failed_chunks
+                        ],
+                    )
+                ],
+            ),
+        ],
+    )
+
+
+def create_result_model(allowed_types: list[str] | None = None) -> type[Result]:
+    """
+    Factory function to create a Result model with a restricted set of vulnerability types.
+
+    Args:
+        allowed_types: A list of strings representing the allowed vulnerability types.
+                       If None or empty, the default Result model with a string type is returned.
+
+    Returns:
+        A Pydantic model class for Result, with ResultProperties.type restricted to an enum
+        if allowed_types is provided.
+    """
+    if not allowed_types:
+        return Result
+
+    # The type annotations here are for pydantic, may take some more digging to get these to work with mypy.
+    VulnTypeEnum = Enum("VulnTypeEnum", {t: t for t in allowed_types})  # type: ignore[misc]
+
+    class RestrictedResultProperties(ResultProperties):
+        type: VulnTypeEnum = Field(  # type: ignore[valid-type,assignment]
+            description="Type of vulnerability (e.g., 'SQL Injection', 'XSS', 'Command Injection', etc.)"
+        )
+
+    class RestrictedResult(Result):
+        properties: RestrictedResultProperties = Field(
+            description="Key/value pairs that provide additional information about the result."
+        )
+
+    return RestrictedResult
+
+
+def create_run_model(allowed_types: list[str] | None = None) -> type[Run]:
+    """
+    Factory function to create a Run model with a restricted set of vulnerability types.
+
+    Args:
+        allowed_types: A list of strings representing the allowed vulnerability types.
+                       If None or empty, the default Run model with a string type is returned.
+
+    Returns:
+        A Pydantic model class for Run, with ResultProperties.type restricted to an enum
+        if allowed_types is provided.
+    """
+    RestrictedResultModel = create_result_model(allowed_types)
+
+    # The type annotations here are for pydantic, may take some more digging to get these to work with mypy.
+    class RestrictedRun(Run):
+        results: List[RestrictedResultModel] = Field(  # type: ignore[valid-type]
+            description="The set of results contained in a SARIF log."
+        )
+
+    return RestrictedRun
```

Comment:
This allows dynamically restricting the vulnerability types which is helpful when determining equality when benchmarking.